### ISSUE TITLE
[Feature Request] Support for Podman

### ISSUE THREAD

**Issue by R8s6** [containrrr/watchtower#1060]
**Is your feature request related to a problem? Please describe.**
Currently `watchtower` requires `/var/run/docker.sock`, which is not present in a system with only `podman` installed (i.e. where `docker` is not installed).

**Describe the solution you'd like**
Potentially supporting `podman` in the future.

Thanks!

**Comment by simonaronsson** [comment]
Looks like podman has a docker-compatible REST api, so should be working out of the box as far as I can tell. I haven't tried it myself though: https://podman.io/blogs/2020/07/01/rest-versioning.html

**Comment by piksel** [comment]
if [podman service](https://docs.podman.io/en/latest/_static/api.html) is running, this should work:
```
podman run -v /var/run/podman/podman.sock:/var/run/docker.sock docker.io/containrrr/watchtower 
```

**Comment by piksel** [comment]
Update; it sort of worked when I tried it:
```
2021-08-31T18:00:53Z [D] Doing a HEAD request to fetch a digest
                         url: https://index.docker.io/v2/containrrr/watchtower/manifests/latest
2021-08-31T18:00:53Z [D] Found a remote digest to compare with
                         remote: sha256:3283e0b5be326d77ff4f4e8b7a91d46aaa1d511c74877b5a32f161548812d00c
2021-08-31T18:00:53Z [D] Comparing
                         local: sha256:3283e0b5be326d77ff4f4e8b7a91d46aaa1d511c74877b5a32f161548812d00c
                         remote: sha256:3283e0b5be326d77ff4f4e8b7a91d46aaa1d511c74877b5a32f161548812d00c
2021-08-31T18:00:53Z [D] Found a match
2021-08-31T18:00:53Z [D] No pull needed. Skipping image.
2021-08-31T18:00:53Z [I] Found new docker.io/containrrr/watchtower:latest image (9167b324e914)
2021-08-31T18:00:53Z [D] This is the watchtower container /sweet_greider
2021-08-31T18:00:53Z [D] Renaming container /sweet_greider (4c4ec40ff1a6) to PPctWJctHvcNrpXpLaCSQaSdAmRtxEdN
2021-08-31T18:00:53Z [I] Creating /sweet_greider
2021-08-31T18:00:53Z [E] Error response from daemon: fill out specgen: ulimit option "RLIMIT_NOFILE=1048576:1048576" requires name=SOFT:HARD, failed to be parsed: invalid ulimit type: RLIMIT_NOFILE
2021-08-31T18:00:53Z [D] Session done: 1 scanned, 0 updated, 1 failed
```

Not really sure what is going on here. It checks the watchtower image and concludes that `No pull needed. Skipping image.`, but then goes ahead and updates it anyway?
Then it fails when trying to rename itself. It might also just be that my podman installation is broken... perhaps someone with a known working setup can try it and share their experience?

**Comment by piksel** [comment]
The cause of the error above is: https://github.com/containers/podman/issues/9803 (Finding 4)
Additionally, the "name" is not accepted by podman, since it has a leading slash:

```
INFO[0003] Creating /nginx-test
ERRO[0003] Error response from daemon: container create: error running container create option: names must match [a-zA-Z0-9][a-zA-Z0-9_.-]*: invalid argument
```

By patching these two in the new config I were able to successfully recreate a podman container:

```go
// pkg/container/client.go:214
hostConfig.Ulimits = nil
name = name[1:]
```

<details>
<summary>Log output from successful run</summary>

```
nils@xiwangmu:~/src/watchtower $ go build && sudo ./watchtower --trace --run-once --host unix:///var/run/podman/podman.sock
DEBU[0000]
DEBU[0000] Sleeping for a second to ensure the docker api client has been properly initialized.
DEBU[0001] Making sure everything is sane before starting
INFO[0001] Watchtower v0.0.0-unknown
Using no notifications
Checking all containers (except explicitly disabled with label)
Running a one time update.
WARN[0001] trace level enabled: log will include sensitive information as credentials and tokens
DEBU[0001] Checking containers for updated images
DEBU[0001] Retrieving running containers
DEBU[0001] Trying to load authentication credentials.    container=/nginx-test image="docker.io/library/nginx:latest"
DEBU[0001] No credentials for docker.io found            config_file=/config.json
DEBU[0001] Got image name: docker.io/library/nginx:latest
DEBU[0001] Checking if pull is needed                    container=/nginx-test image="docker.io/library/nginx:latest"
DEBU[0001] Building challenge URL                        URL="https://index.docker.io/v2/"
DEBU[0001] Got response to challenge request             header="Bearer realm=\"https://auth.docker.io/token\",service=\"registry.docker.io\"" status="401 Unauthorized"
DEBU[0001] Checking challenge header content             realm="https://auth.docker.io/token" service=registry.docker.io
DEBU[0001] Setting scope for auth token                  image=docker.io/library/nginx scope="repository:library/nginx:pull"
DEBU[0001] No credentials found.
DEBU[0002] Parsing image ref                             host=index.docker.io image=docker.io/library/nginx normalized="docker.io/library/nginx:latest" tag=latest
TRAC[0002] Setting request token                         
DEBU[0002] Doing a HEAD request to fetch a digest        url="https://index.docker.io/v2/library/nginx/manifests/latest"
DEBU[0002] Found a remote digest to compare with         remote="sha256:4d4d96ac750af48c6a551d757c1cbfc071692309b491b70b2b8976e102dd3fef"
DEBU[0002] Comparing                                     local="sha256:4d4d96ac750af48c6a551d757c1cbfc071692309b491b70b2b8976e102dd3fef" remote="sha256:4d4d96ac750af48c6a551d757c1cbfc071692309b491b70b2b8976e102dd3fef"
DEBU[0002] Found a match
DEBU[0002] No pull needed. Skipping image.
INFO[0002] Found new docker.io/library/nginx:latest image (a25dfb1cd178)
INFO[0002] Stopping /nginx-test (5008385716af) with SIGTERM
DEBU[0011] Removing container 5008385716af
INFO[0011] Creating nginx-test
DEBU[0011] Starting container /nginx-test (4a74f0351cfb)
DEBU[0011] Session done: 1 scanned, 0 updated, 0 failed
Waiting for the notification goroutine to finish
```

</details>

This still leaves the issue of it always trying to update the containers though:
```
DEBU[0002] No pull needed. Skipping image.
INFO[0002] Found new docker.io/library/nginx:latest image (a25dfb1cd178)
```

**Comment by piksel** [comment]
Found the cause of the image always being treated as "stale":
```
nils@xiwangmu:~ $ sudo curl -s --unix-socket /run/podman/podman.sock 'http://d/v3.0.0/containers/nginx-test/json' | jq -C '.Image'
"docker.io/library/nginx:latest"
```
Podman doesn't return the image ID in the container inspect result, giving the image "name" instead.
This could be solved by using the podman-specific API endpoint:
```
sudo curl -s --unix-socket /run/podman/podman.sock 'http://d/v3.0.0/libpod/containers/nginx-test/json' | jq -C .Image
"a25dfb1cd178de4f942ab5a87d3d999e3f981bb9f36fc6ee38b04669e14c32d2"
```

So, overall, this **could** be implemented, but would require a special flag for "podman mode".

**Comment by ghost** [comment]
+1 I'm considering transitioning future systems to use Podman and would like to see these edge cases worked out.

**Comment by simonaronsson** [comment]
> +1 I'm considering transitioning future systems to use Podman and would like to see these edge cases worked out.

feel free to join in on the efforts 👍🏽

**Comment by dbrgn** [comment]
Note that podman has something like watchtower built-in: <https://docs.podman.io/en/latest/markdown/podman-auto-update.1.html>

**Comment by R8s6** [comment]
> Note that podman has something like watchtower built-in: https://docs.podman.io/en/latest/markdown/podman-auto-update.1.html

I see one needs to create a systemd unit for the container needed to be auto-updated. If I were to have all my containers auto-updated, does it mean I need to create a systemd unit file for each one (instead of allowing the systemd unit to auto-update all containers, as well as those to be installed in the future)? Thanks

**Comment by d-513** [comment]
@piksel it now returns the image id, so this could be done. ULimits still an issue.

**Comment by MartinX3** [comment]
@R8s6 please read the linked docs.
Podman generates the systemd files for you.

**Comment by d-513** [comment]
I made a PR, but still facing this issue:
```
cannot set memory swappiness with cgroupv2: OCI runtime error
```
Seems that podman does NOT like explicitly specifying everything, even if it was defaulted to that value.

**Comment by DocMAX** [comment]
> I made a PR, but still facing this issue:
> 
> ```
> cannot set memory swappiness with cgroupv2: OCI runtime error
> ```
> 
> Seems that podman does NOT like explicitly specifying everything, even if it was defaulted to that value.

I have this error on multiple containers now. How can i fix what watchtower messed up?

**Comment by d-513** [comment]
> > I made a PR, but still facing this issue:
> > ```
> > cannot set memory swappiness with cgroupv2: OCI runtime error
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Seems that podman does NOT like explicitly specifying everything, even if it was defaulted to that value.
> 
> I have this error on multiple containers now. How can i fix what watchtower messed up?

remove the containers and create them again

**Comment by xtexx** [comment]
The same error.

**Comment by xpahos** [comment]
> The same error.

Try to build podman with this commit: https://github.com/containers/podman/commit/6ea703b79880c7f5119fe5355074f8e971df6626

```
CONTAINER ID  IMAGE                                   COMMAND               CREATED       STATUS       PORTS       NAMES
2e9e8b635337  quay.io/outline/shadowbox:stable        /bin/sh -c /cmd.s...  39 hours ago  Up 39 hours              shadowbox
58d75a814a6a  docker.io/containrrr/watchtower:latest  --cleanup --label...  39 hours ago  Up 39 hours              watchtower
```

**Comment by xtexx** [comment]
thanks!

**Comment by xtexx** [comment]
one container could not be started after re-creation
```
time="2023-05-18T18:10:24Z" level=info msg="Found new docker.io/binwiederhier/ntfy:latest image (2434c49f7c33)"

time="2023-05-18T18:10:26Z" level=info msg="Stopping /ntfy (09b65f2ca3c5) with SIGTERM"

time="2023-05-18T18:10:27Z" level=info msg="Creating /ntfy"

time="2023-05-18T18:10:27Z" level=error msg="Error response from daemon: crun: cannot set memory swappiness with cgroupv2: OCI runtime error"

time="2023-05-18T18:10:27Z" level=info msg="Session done" Failed=1 Scanned=9 Updated=0 notify=no

time="2023-05-18T18:10:27Z" level=error msg="Failed to send shoutrrr notification" error="failed to send ntfy notification: got HTTP 502 Bad Gateway" index=0 notify=no service=ntfy
```

**Comment by MastaG** [comment]
Same here.. it re-creates the container but it fails to start...

**Comment by n-hass** [comment]
> I made a PR, but still facing this issue:
> 
> ```
> 
> cannot set memory swappiness with cgroupv2: OCI runtime error
> 
> ```
> 
> Seems that podman does NOT like explicitly specifying everything, even if it was defaulted to that value.

Confirming issue is still present in podman release v4.6.0

**Comment by soyelmismo** [comment]
and what if no systemd? :p

**Comment by loeffelpan** [comment]
Still present in 4.8.x.
I'm using Nextcloud-AIO with podman rootless. For mastercontainer update it uses watchtower which works fine for docker but not for podman. Getting the swapiness-error on container start after watchtower created the new one.

Does watchtower use something similar like executing `podman container clone ...` via CLI?
I found this one related to the clone command.
https://github.com/containers/podman/issues/13916

Unfortunately it is unsolved. Maybe someone can raise a new issue to solve this in podman?

**Comment by lazyzyf** [comment]
I have the same issue.

**Comment by chisaato** [comment]
> > I made a PR, but still facing this issue:
> > ```
> > 
> > cannot set memory swappiness with cgroupv2: OCI runtime error
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Seems that podman does NOT like explicitly specifying everything, even if it was defaulted to that value.
> 
> Confirming issue is still present in podman release v4.6.0

Same on podman 5.2.2

![image](https://github.com/user-attachments/assets/a9873348-58f0-4e32-b1af-4911177e8fae)

I think this maybe watchtower set memory swappiness to 0 for containers. I do some research abount the low-level runtime of podman used by default, which is crun.  See here

https://github.com/containers/crun/blob/f54d383e0012a698262ca7b93beb971f99b1f238/src/libcrun/cgroup-resources.c#L878

```c
  if (memory->swappiness_present)
    {
      if (cgroup2)
        return crun_make_error (err, 0, "cannot set memory swappiness with cgroupv2");

      len = sprintf (fmt_buf, "%" PRIu64, memory->swappiness);
      ret = write_cgroup_file (dirfd, "memory.swappiness", fmt_buf, len, err);
      if (UNLIKELY (ret < 0))
        return ret;
    }
```
I use `execsnoop` from `bcc-tools` to check what crun does when restaring containers. 

![image](https://github.com/user-attachments/assets/3792ceeb-330c-4465-829e-501a0ff43ea3)
  
Cool,let's dig deeper to this `userdata`

![image](https://github.com/user-attachments/assets/062c634b-88f0-4c44-893d-141edf5e9054)

then I found that containers has set memory.swappiness in `config.json`, you can find the file using 

```bash
podman inspect <container>| grep OCIConfigPath
```

then compare the file with some containers created by podman

![image](https://github.com/user-attachments/assets/f0b19595-a58c-478f-8a11-fba9c89c50d3)

But who gives the 0 value? I inspect a container in a server running with docker, we will see

![image](https://github.com/user-attachments/assets/ecaa8cdc-9a62-4ee5-ab4e-b1a7199b12c7)

How about using podman? 

![image](https://github.com/user-attachments/assets/3bccafc5-149f-4f29-b8bd-e38962c94c40)

well, podman gives zero even this value is not set. Then watchtower copy this and send container info to docker socket(emulated by podman) with zero setted in `swappiness` and BOOOM! , crun rejects it on CGroupV2.

**Comment by tuxillo** [comment]
What's missing here to get the same functionality as for docker?

**Comment by chisaato** [comment]
> What's missing here to get the same functionality as for docker?

podman needs to return same value as docker api. and podman says it maybe fix in v6

**Comment by lazyzyf** [comment]
> > What's missing here to get the same functionality as for docker?
> 
> podman needs to return same value as docker api. and podman says it maybe fix in v6

so this feature is gated by podman, rather than watchtower, right?

**Comment by warlordattack** [comment]
This compose file works with Rocky linux (proxmox VM) / Podman : 

1. disable SELinux in Rocky linux
2. in the compose file do not use : user: ${PUID}:${PGID}
3. add : 
    security_opt: 
      - label=disable

docker.sock is automatically translated to podman.sock by podman, so this mapping works in my Podman : 
- /var/run/docker.sock:/var/run/docker.sock:rw

```
---
services:
  srv_watchtower:
    container_name: ${C_WT}
    hostname: ${C_WT_HOST}
    image: ${C_WT_IMG}
    restart: ${C_ALL_RESTART}
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock:rw # check 1. 2. 3. !!!
    # 1. - user: ${PUID}:${PGID} # UID/GID disabled
    security_opt: # 2. SELinux disabled
      - label=disable # 3. - label=disable added
    environment:
      - TZ=${TZ}
      #- WATCHTOWER_RUN_ONCE=true
      - WATCHTOWER_POLL_INTERVAL=86400 # 1h=3600 # 24h=86400
      - WATCHTOWER_MONITOR_ONLY=true
      - WATCHTOWER_CLEANUP=true
      #- WATCHTOWER_ROLLING_RESTART=true # restart containers one by one
      - WATCHTOWER_REVIVE_STOPPED=false
      - WATCHTOWER_INCLUDE_STOPPED=true
      #- WATCHTOWER_DEBUG=true      
      #- WATCHTOWER_LABEL_ENABLE=${WATCHTOWER_LABEL_ENABLE}
      - WATCHTOWER_NOTIFICATIONS=gotify
      - WATCHTOWER_NOTIFICATION_GOTIFY_URL=http://${C_GOT_HOST}:${C_GOT_80}/
      - WATCHTOWER_NOTIFICATION_GOTIFY_TOKEN=${C_WT_GOTIFY}
    labels:
      - "com.centurylinklabs.watchtower.enable=${WATCHTOWER_LABEL_ENABLE}"
    ports:
      - ${C_WT_80}:80
      - ${C_WT_443}:443
    networks:
      web:
        ipv4_address: ${C_WT_IP_WEB}
networks: 
  web: 
    external: true
```

![image](https://github.com/user-attachments/assets/df0bbbbc-3270-4dfd-9cf0-c0ec1955d0b3)

Best regards :)

Had this output after run (don't know if update errors are due to podman) : 

```
Unable to update container "/r-koodo": Error response from daemon: {"message":"x509: certificate is valid for 6b43d87c96397f68a135c01c1f292377.9058f1fa15a681094bc241d87108e81b.traefik.default, not localhost"}. Proceeding to next.
Could not do a head request for "docker.io/matrixdotorg/synapse:latest", falling back to regular pull.
Reason: Get "https://index.docker.io/v2/": dial tcp: lookup index.docker.io on 192.168.10.1:53: read udp 192.168.10.86:36764->192.168.10.1:53: i/o timeout
Unable to update container "/net-perplexica-b": Error response from daemon: {"message":"x509: certificate is valid for 6b43d87c96397f68a135c01c1f292377.9058f1fa15a681094bc241d87108e81b.traefik.default, not localhost"}. Proceeding to next.
Unable to update container "/net-perplexica-f": Error response from daemon: {"message":"x509: certificate is valid for 6b43d87c96397f68a135c01c1f292377.9058f1fa15a681094bc241d87108e81b.traefik.default, not localhost"}. Proceeding to next.
Found new docker.io/rustdesk/rustdesk-server:latest image (e0892e67d5a7)
Could not do a head request for "docker.io/tensorchord/pgvecto-rs@sha256:90724186f0a3517cf6914295b5ab410db9ce23190a2d9d0b9dd6463e3fa298f0", falling back to regular pull.
Reason: Parsed container image ref has no tag: docker.io/tensorchord/pgvecto-rs@sha256:90724186f0a3517cf6914295b5ab410db9ce23190a2d9d0b9dd6463e3fa298f0
Found new ghcr.io/immich-app/immich-server:release image (fd58d779b0a6)
Found new docker.io/tiredofit/traefik-cloudflare-companion:latest image (e620103b4344)
Found new docker.io/rustdesk/rustdesk-server:latest image (e0892e67d5a7)
Found new ghcr.io/immich-app/immich-server:release image (fd58d779b0a6)
```

And this is my portainer compose file witch also uses docker.sock with podman/rocky and works : 

```
---
services:
  srv_portainer:
    container_name: ${C_PTN}
    hostname: ${C_PTN_HOST}
    image: ${C_PTN_IMG}
    restart: ${C_ALL_RESTART}
    ports:
      - ${C_PTN_9000}:9000
      - ${C_PTN_8000}:8000
      - ${C_PTN_9443}:9443
    volumes:
      - ${REP_SSL}:/certs
      - ${REP_APPDATA}/${C_PTN}/data:/data:rw
      - ${REP_APPDATA}/${C_PTN}/config:/config:rw
      - /etc/localtime:/etc/localtime:ro
      - /var/run/docker.sock:/var/run/docker.sock:rw # check 1. 2. 3. !!!
    # 1. - user: ${PUID}:${PGID} # UID/GID disabled
    security_opt: # 2. SELinux disabled
      - label=disable # 3. - label=disable added
    environment:
      - PUID=${PUID}
      - PGID=${PGID}
      - UMASK=${UMASK}
      - TZ=${TZ}
      - AGENT_SECRET=${C_PTN_AGENT_SECRET}
    labels:
      - "com.centurylinklabs.watchtower.enable=${WATCHTOWER_LABEL_ENABLE}"
    networks:
      web:
        ipv4_address: ${C_PTN_IP_WEB}
networks: 
  web: 
    external: true
```

**Comment by lazyzyf** [comment]
how come you map the volume to /var/run/docker.sock?
mine is /run/user/1000/podman/podman.sock.

**Comment by warlordattack** [comment]
The mapping (full docker compose in my other post) : 
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock:rw 
    security_opt: 
      - label=disable 

you need to :
1. disable SELINUX in Rocky Linux (find tutorial in google to do this)
2. remove "user:" in your docker compose 
3. add security_opt: -label=disable in your docker compose

Podaman translates docker directives to podman directives automaticly (you do not need use podman.sock, use docker.sock instead).
Try my docker compose as presented in other post.

![image](https://github.com/user-attachments/assets/2ba553c0-d5f8-4cfc-b565-673ed51a7903)

**Comment by jiahaojz-onum** [comment]
Would it be possible to add a flag to Watchtower to avoid passing the MemorySwappiness to the newly created container? AFAIK, the issue is that podman defines MemorySwappiness to 0 by default as @chisaato checked, and when it calls the socket to recreate the container I guess it passes this parameter which leads up to the issue we are facing.

```
podman run -d --name nginx nginx
70636db2ce09471407a568d9ab6e3e84476e2e800f7571c0413586521c4eacb5

podman inspect nginx | grep -i swappiness
               "MemorySwappiness": 0,

podman run --memory-swappiness=0 -d --name nginx-2 nginx
Error: crun: cannot set memory swappiness with cgroupv2: OCI runtime error

```

The reason I suggest this solution is that if I am not wrong MemorySwappiness is still present due to retrocompatibilty with cgroupsv1, setting a flag in watchtower to not passing that MemorySwappiness would not introduce any breaking changes and would allow solving this issue

**Comment by chisaato** [comment]
consider to compatibility, maybe we should check more fields.

**Comment by jiahaojz-onum** [comment]
Okey, definitely this requires more research, but it is a first approach, as context I have no experience contributing in open source projects and I am new to golang. But I have tried to set here https://github.com/containrrr/watchtower/blob/main/pkg/container/container.go#L354, which is the part where watchtower creates the hostConfig for the container the following value

`hostConfig.MemorySwappiness = nil`

And managed to make watchtower to work with podman, maybe we could use a flag like --reset-memory-swappiness-on-creation to enable that configuration. After that if we inspect the recreated container, podman sets the MemorySwappiness to 0 again, but the hostConfig was set to nil prior to the container creation avoiding the issue `Error: crun: cannot set memory swappiness with cgroupv2: OCI runtime error`

```
podman inspect webserver | grep -i swappiness
               "MemorySwappiness": 0,
```

Do you think we could make it work somehow going down this path? After all it seems we cannot set MemorySwappiness anyway, and with the flag we would only enable this behaviour if required. Also I think it could be quicker than waiting for Podman v6

**Comment by jiahaojz-onum** [comment]
This is an example (A lot of work to do yet, first approach to allow the compatibility with podman):

I have added some debugging

![image](https://github.com/user-attachments/assets/b92359a9-3c05-4ca0-945f-58fc3aea1836)

Then I execute watchtower (We get the default and expected behaviour)

![image](https://github.com/user-attachments/assets/18889852-1df2-45ad-83f8-6adca7b817c3)

The I add the removal of the MemorySwappiness

![image](https://github.com/user-attachments/assets/ab1fc517-3035-4964-9515-2902d96ca732)

And execute again

![image](https://github.com/user-attachments/assets/9ea03978-2f8e-4572-ae13-705fc4c050c3)

Now it succeeds

**Comment by soyji** [comment]
Hi @piksel or someone, could you review this? https://github.com/containrrr/watchtower/pull/2072, jiahaojz-onum is my account too, so I am following that approach with the flag.

Unfortunately for the people in the thread I am seeing that the latest release was in November 2023...

**Comment by apparle** [comment]
> Unfortunately for the people in the thread I am seeing that the latest release was in November 2023...

Given there's no activity on this repo, I'm guessing this is project is not being actively maintained.

Because I need watchtower as Nextcloud AIO uses it under-the-hood, I figured out a workaround to get podman (v4.9.3 on Ubuntu 24.04) working:
1. I took @soyji 's PR https://github.com/containrrr/watchtower/pull/2072 , forked it and disabled memory swappiness by default. My branch is here: https://github.com/apparle/watchtower/tree/podman-compatibility . 
2. Next I built this as a docker image and pushed it to docker hub: https://hub.docker.com/repository/docker/apparle/watchtower/general 
3. I tricked podman into pulling my image for the tag `containrrr/watchtower:latest` by creating `~/.config/containers/registries.conf` with below contents:
    ```
    [[registry]]
    prefix="docker.io/containrrr/watchtower"
    location = "docker.io/apparle/watchtower"
    ```

Hopefully this helps someone else too - you can probably just do step-3 and use my image.

**Comment by AndydeCleyre** [comment]
In case this is helpful news to anyone, I think similar behavior to watchtower can be achieved with systemd + [podlet](https://github.com/containers/podlet) + [quadlet](https://docs.podman.io/en/stable/markdown/podman-systemd.unit.5.html) + podman-auto-update.timer .

**Comment by GingerGraham** [comment]
> In case this is helpful news to anyone, I think similar behavior to watchtower can be achieved with systemd + [podlet](https://github.com/containers/podlet) + [quadlet](https://docs.podman.io/en/stable/markdown/podman-systemd.unit.5.html) + podman-auto-update.timer .

This is the route I ended up going down. A bit more work to get setup, but can confirm this works.

**Comment by Tobi-De** [comment]
> > In case this is helpful news to anyone, I think similar behavior to watchtower can be achieved with systemd + [podlet](https://github.com/containers/podlet) + [quadlet](https://docs.podman.io/en/stable/markdown/podman-systemd.unit.5.html) + podman-auto-update.timer .
> 
> This is the route I ended up going down. A bit more work to get setup, but can confirm this works.

Is there some kind of easy to follow tutorial on this out there ?

**Comment by AndydeCleyre** [comment]
> Is there some kind of easy to follow tutorial on this out there ?

I'll try, and hopefully someone can correct if I'm wrong.

- Install podlet.
- Create the container units folder. If root, that's `/etc/containers/systemd/`. Otherwise it's `$XDG_CONFIG_HOME/containers/systemd/` (generally `~/.config/containers/systemd/`).
- Whatever `podman` command you use to start your container, prepend it with `podlet -i -u` to generate a `something.container` file in the units folder from the last step. See the [podlet readme](https://github.com/containers/podlet) for more advanced usage.
- Edit the generated file to add `AutoUpdate=registry` to the `[Container]` section.
- Ensure the service is discovered/generated, and start it:

```console
$ systemctl --user daemon-reload
$ systemctl --user start something
```

Replace "something" with the generated filename, without the `.container` suffix. Don't use `--user` if using podman as root.
- Ensure the auto-updating timer is enabled:

```console
$ systemctl --user enable --now podman-auto-update.timer
```

Again, only use `--user` if appropriate.

**Comment by aanno** [comment]
Is this still relevant?

Podman has evolved substantially over the years. And it certainly supports the docker engine API out of the box, even in rootless mode. The following lines start the socket:

```bash
# start socket in rootless mode
systemctl --user restart podman.socket
# check if socket is up
curl -H "Content-Type: application/json" \
        --unix-socket /run/user/$UID/podman/podman.sock \
    http://localhost/_ping
export DOCKER_SOCKET=/run/user/$UID/podman/podman.sock
export DOCKER_HOST=unix://$DOCKER_SOCKET
echo "export DOCKER_SOCKET=$DOCKER_SOCKET"
echo "export DOCKER_HOST=$DOCKER_HOST"
```

The socket is not at `/var/run/docker.sock`, but at `/run/user/$UID/podman/podman.sock`. But apart from that it is up and kicking. In my experiments the watchtower container has not been a problem.

**Comment by GingerGraham** [comment]
> Is this still relevant?
> 
> Podman has evolved substantially over the years. And it certainly supports the docker engine API out of the box, even in rootless mode. The following lines start the socket:
> 
> # start socket in rootless mode
> systemctl --user restart podman.socket
> # check if socket is up
> curl -H "Content-Type: application/json" \
>         --unix-socket /run/user/$UID/podman/podman.sock \
>     http://localhost/_ping
> export DOCKER_SOCKET=/run/user/$UID/podman/podman.sock
> export DOCKER_HOST=unix://$DOCKER_SOCKET
> echo "export DOCKER_SOCKET=$DOCKER_SOCKET"
> echo "export DOCKER_HOST=$DOCKER_HOST"
> The socket is not at `/var/run/docker.sock`, but at `/run/user/$UID/podman/podman.sock`. But apart from that it is up and kicking. In my experiments the watchtower container has not been a problem.

I've done something similar as a workaround and it seems to work fine.

My only comment would be, this is a workaround and it would be better to see watchtower detect the podman socket itself rather than needing to be worked around.

**Comment by aanno** [comment]
Well, I found out the following:

* it is still relevant, see https://github.com/nextcloud/all-in-one/discussions/5994#discussioncomment-12738550 for details
* nextcloud/all-in-one is moving away from this version of watchtower to https://github.com/nicholas-fedor/watchtower/discussions/267#discussioncomment-13201594, hence it might not be a problem with version v11.2.0 and up, see https://github.com/nextcloud/all-in-one/pull/6533 for details

**Comment by oilervoss** [comment]
I want to highlight the very well maintened [nickfedor fork](https://github.com/nicholas-fedor/watchtower/) here: https://watchtower.nickfedor.com
I just added to my previous compose:
```
    environment:
      WATCHTOWER_DISABLE_MEMORY_SWAPPINESS: true
      WATCHTOWER_CPU_COPY_MODE: auto
```
and changed:
```
    volumes:
      - /run/podman/podman.sock:/var/run/docker.sock
```

I also added `      WATCHTOWER_UPDATE_ON_START: true`. I believe it's a nice nick's addition.

**Comment by lazyzyf** [comment]
> I want to highlight the very well maintened [nickfedor fork](https://github.com/nicholas-fedor/watchtower/) here: https://watchtower.nickfedor.com I just added to my previous compose:
> 
> ```
>     environment:
>       WATCHTOWER_DISABLE_MEMORY_SWAPPINESS: true
>       WATCHTOWER_CPU_COPY_MODE: auto
> ```
> 
> and changed:
> 
> ```
>     volumes:
>       - /run/podman/podman.sock:/var/run/docker.sock
> ```
> 
> I also added ` WATCHTOWER_UPDATE_ON_START: true`. I believe it's a nice nick's addition.

You are running podman with root or rootless ?

**Comment by oilervoss** [comment]
> I want to highlight the very well maintened [nickfedor fork](https://github.com/nicholas-fedor/watchtower/) here: https://watchtower.nickfedor.com I just added to my previous compose:
> 
> ```
>     environment:
>       WATCHTOWER_DISABLE_MEMORY_SWAPPINESS: true
>       WATCHTOWER_CPU_COPY_MODE: auto
> ```
> 
> and changed:
> 
> ```
>     volumes:
>       - /run/podman/podman.sock:/var/run/docker.sock
> ```
> 
> I also added ` WATCHTOWER_UPDATE_ON_START: true`. I believe it's a nice nick's addition.

Running podman rootful. For those who are running rootless, I know of one necessary change:

```
    volumes:
      - /run/user/1000/podman/podman.sock:/var/run/docker.sock  #depending on your podman installation. you need to find what is your podman.sock path and replace the number 1000 here for whatever you find.
```

**Comment by lazyzyf** [comment]
> > I want to highlight the very well maintened [nickfedor fork](https://github.com/nicholas-fedor/watchtower/) here: https://watchtower.nickfedor.com I just added to my previous compose:
> > 
> > ```
> >     environment:
> >       WATCHTOWER_DISABLE_MEMORY_SWAPPINESS: true
> >       WATCHTOWER_CPU_COPY_MODE: auto
> > ```
> > 
> > and changed:
> > 
> > ```
> >     volumes:
> >       - /run/podman/podman.sock:/var/run/docker.sock
> > ```
> > 
> > I also added ` WATCHTOWER_UPDATE_ON_START: true`. I believe it's a nice nick's addition.
> 
> Running podman rootful. For those who are running rootless, I know of one necessary change:
> 
> ```
>     volumes:
>       - /run/user/1000/podman/podman.sock:/var/run/docker.sock  #depending on your podman installation. you need to find what is your podman.sock path and replace the number 1000 here for whatever you find.
> ```

 thanks, it is working now.
